### PR TITLE
Feature/random forests

### PR DESCRIPTION
**Pull request by cedric-len**
The Decision Tree branch has to be merged first!

### CODE REVIEW

File: src/GiniImpurity.jl [hunk 1]
```diff
@@ -77,12 +77,16 @@ Args:
...
-function find_best_split(data::Matrix{T}, labels::Vector{L}) where {T, L}  
+function find_best_split(data::Matrix{T}, labels::Vector{L}, num_features_to_use::Int=-1) where {T, L}  
     best_gini = Inf  
     best_feature_index = -1  
     best_feature_value = -1  
     num_features = size(data, 2)
-    for feature_index in 1:num_features  
+    features_to_use = 1:num_features
+    if (num_features_to_use != -1)
+        features_to_use = sample(1:num_features, num_features_to_use, replace=true)
```
Comment (Enea09) [inline src/GiniImpurity.jl]: shouldn't the choice of features here be without replacement?

File: src/RandomForest.jl [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+Represents a RandomForest.
+
+`trees` is the vector of ClassificationTree structures.
+`num_features` is the number of features to use when finding the best split. If -1, all the features are used.
+"""
+struct RandomForest{T, L}
+    trees::Vector{ClassificationTree{T, L}}
+    num_features::Int
+
+    function RandomForest(data::Matrix{T}, labels::Vector{L}, max_depth::Int, min_samples_split::Int ,number_of_trees::Int, subsample_percentage::Float64, num_features::Int) where {T, L}
```
Comment (Enea09) [inline src/RandomForest.jl]: I would suggest moving the construcion and training of the trees to the fit function and including the data and labels as parameters in that function

File: src/GiniImpurity.jl (hunk 1 above)
Comment (cedric-len) [reply src/GiniImpurity.jl]: I am unsure. Is there some universally valid implementation of the algorithm?

File: src/RandomForest.jl (hunk 2 above)
Comment (cedric-len) [reply src/RandomForest.jl]: I agree. But to minimize merge conflicts. I would do the refactoring after we merged the regression and forest feature together :)

File: src/GiniImpurity.jl (hunk 1 above)
Comment (Enea09) [reply src/GiniImpurity.jl]: I think if we select features with replacements we might select the same feature more than once
So if num_features_to_use is 3 we actually might use the information from only 1/2 features when finding the best split. Some examples I saw online used it without replacements as well
For example: https://youtu.be/NxEHSAfFlK8?si=jtKwL2SZRg8DQUQo&t=996

Comment (cedric-len) [reply src/GiniImpurity.jl]: Then i will change it

**Pull request merged**